### PR TITLE
OSD-13509: AVO support to create private zones

### PR DESCRIPTION
**Pull request by robotmaxtron**
New field added to the VPCE spec to support AVO's creation and tagging of route53 additional hosted private zones.

### CODE REVIEW

**Comment by robotmaxtron** [comment]
@tonytheleg an example has been added

File: pkg/aws_client/tags.go [hunk 1]
```diff
@@ -18,11 +18,21 @@ package aws_client
...
 func (c *AWSClient) CreateTags(ctx context.Context, input *ec2.CreateTagsInput) (*ec2.CreateTagsOutput, error) {
 	return c.ec2Client.CreateTags(ctx, input)
 }
+
+// ListTagsForResource will fetch tags of a hosted zone or healthcheck
+func (c *AWSClient) ListTagsForResource(ctx context.Context, params *route53.ListTagsForResourceInput) (*route53.ListTagsForResourceOutput, error) {
+	return c.route53Client.ListTagsForResource(ctx, params)
+}
+
+// ChangeTagsForResource adds, edits, or deletes tags for a health check or a hosted zone.
```
Comment (mjlshen) [inline pkg/aws_client/tags.go:35]: Based on the underlying implementation I think this only adds tags

File: pkg/aws_client/route53_hosted_zone.go [hunk 2]
```diff
@@ -96,3 +96,44 @@ func (c *AWSClient) DeleteResourceRecordSet(ctx context.Context, rrs *types.Reso
...
+// GeneratePrivateHostedZoneInput generates the required input for a PrivateHostedZone
+func (c *AWSClient) GeneratePrivateHostedZoneInput(domain, vpcId, callerRef, region string) *route53.CreateHostedZoneInput {
+	zoneInput := &route53.CreateHostedZoneInput{
+		CallerReference:  aws.String(callerRef),
+		Name:             aws.String(domain),
+		HostedZoneConfig: &types.HostedZoneConfig{Comment: aws.String("Created via AVO"), PrivateZone: true},
+		VPC:              &types.VPC{VPCId: aws.String(vpcId), VPCRegion: types.VPCRegion(region)},
+	}
+	return zoneInput
+}
```
Comment (mjlshen) [inline pkg/aws_client/route53_hosted_zone.go:118]: Should these be one function?

File: dev/vpce_hosted_zone_example.yml [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+  subdomainName: "splunk"
+  serviceName: "com.amazonaws.vpce.us-east-1.vpce-svc-00000000000000000"
+  securityGroup:
+    ingressRules:
+      - fromPort: 9997
+        toPort: 9997
+        protocol: tcp
+  externalNameService:
+    name: examplesvcname
+addtlHostedZone: "sometestingdoma.in"
```
Comment (mjlshen) [inline dev/vpce_hosted_zone_example.yml:17]:
```suggestion
  addtlHostedZone: "sometestingdoma.in"
```

File: controllers/vpcendpoint/helpers.go [hunk 4]
```diff
@@ -561,3 +560,39 @@ func tagsContains(tags []ec2Types.Tag, tagsToCheck map[string]string) bool {
...
+				break
+			}
+		}
+		if !found { //Apply Missing Tags
+			_, err := r.awsClient.ChangeTagsForResource(ctx, generatedDefaultTagInput)
+			if err != nil {
+				return fmt.Errorf("failed tag hosted zone with default tags %w", err)
+			}
+		}
+	}
```
Comment (mjlshen) [inline controllers/vpcendpoint/helpers.go:596]: It's a little more straightforward to assemble a map of the actual tags to compare a slice of default tags against, something like (if we make an `actualTagsMap`):

```go
for _, tag := range defaultTags {
	v, ok := actualTagsMap[tag.key]
  // If doesn't exist or the value doesn't match the expected default tag
	if !ok || v != tag.Value {
	  if _, err := r.awsClient.ChangeTagsForResource(ctx, generatedDefaultTagInput); err != nil {
		return fmt.Errorf("failed tag hosted zone with default tags %w", err)
	}
}
```

File: controllers/vpcendpoint/validation.go [hunk 5]
```diff
@@ -274,3 +276,48 @@ func (r *VpcEndpointReconciler) validateExternalNameService(ctx context.Context,
...
 }
+
+// validatePrivateHostedZone attempts to find if the AddtlHostedZoneName has a zone, and if that zone is private, creating it if not found.
+func (r *VpcEndpointReconciler) validatePrivateHostedZone(ctx context.Context, resource *avov1alpha1.VpcEndpoint) error {
+	// AddtlHostedZoneName is optional, return if not present
+	if resource.Spec.AddtlHostedZoneName == "" {
+		return nil
+	}
+	//TODO: Error handling during private zone lookup
+	zoneOut, err := r.awsClient.GetDefaultPrivateHostedZoneId(ctx, resource.Spec.AddtlHostedZoneName)
```
Comment (mjlshen) [inline controllers/vpcendpoint/validation.go:287]: At minimum this, friendlier error message always welcome :D
```suggestion
	zoneOut, err := r.awsClient.GetDefaultPrivateHostedZoneId(ctx, resource.Spec.AddtlHostedZoneName)
	if err != nil {
	  return err
	}
```

File: controllers/vpcendpoint/validation.go [hunk 6]
```diff
@@ -274,3 +276,48 @@ func (r *VpcEndpointReconciler) validateExternalNameService(ctx context.Context,
...
+	trimmedZoneID := strings.TrimPrefix(aws.ToString(zoneOut.Id), "/hostedzone/")
+
+	// TODO: Error handling for edge cases when looking up private zone
```
Comment (mjlshen) [inline controllers/vpcendpoint/validation.go:290]: Is this TODO still relevant? Not sure what it refers to

**Review by mjlshen — CHANGES_REQUESTED** [review]
Mostly seems reasonable, since it's optional there are no conflicts with HyperShift

File: api/v1alpha1/vpcendpoint_types.go [hunk 7]
```diff
@@ -64,13 +64,18 @@ type VpcEndpointSpec struct {
 
 	// ExternalNameService configures the name and namespace of the created Kubernetes ExternalName Service
 	ExternalNameService ExternalNameServiceSpec `json:"externalNameService"`
+
+	// AddtlHostedZoneName is a FQDN to support supplemental VPCE routing
```
[the remark this answers is not in the document]
Comment (robotmaxtron) [reply api/v1alpha1/vpcendpoint_types.go:68]: comment updated

File: pkg/aws_client/route53_hosted_zone.go [hunk 8]
```diff
@@ -96,3 +96,44 @@ func (c *AWSClient) DeleteResourceRecordSet(ctx context.Context, rrs *types.Reso
...
+
+// GenerateDefaultTagsForHostedZoneInput generates the ChangeTagsForResourceInput using the default tags for the zoneId
+func (c *AWSClient) GenerateDefaultTagsForHostedZoneInput(zoneId, clusterTagKey string) (*route53.ChangeTagsForResourceInput, error) {
+	defaultTags, err := util.GenerateR53Tags(clusterTagKey)
+
+	changeTagsInput := &route53.ChangeTagsForResourceInput{
+		ResourceId:    aws.String(zoneId),
+		ResourceType:  types.TagResourceTypeHostedzone,
+		AddTags:       defaultTags,
```
[the remark this answers is not in the document]
Comment (robotmaxtron) [reply pkg/aws_client/route53_hosted_zone.go:127]: Tested this by editing one of the required tag pairs, and removed the second pair completely, the reconciler replaced both the edited and missing tag pairs.

File: dev/vpce_hosted_zone_example.yml (hunk 3 above)
Comment (robotmaxtron) [reply dev/vpce_hosted_zone_example.yml:17]: Updated, also fixed field name.

**Comment by robotmaxtron** [comment]
@mjlshen Thanks for the review, working on an update that includes your suggestions.

File: pkg/aws_client/route53_hosted_zone.go (hunk 2 above)
Comment (robotmaxtron) [reply pkg/aws_client/route53_hosted_zone.go:118]: Updated to a single function

File: controllers/vpcendpoint/helpers.go (hunk 4 above)
Comment (robotmaxtron) [reply controllers/vpcendpoint/helpers.go:596]: Thank you, this has been implemented.

File: controllers/vpcendpoint/validation.go (hunk 6 above)
Comment (robotmaxtron) [reply controllers/vpcendpoint/validation.go]: Removed

**Review by mjlshen — APPROVED** [review]
/lgtm

**Pull request merged**